tests/paramgrill.c

```diff
@@ -632,131 +1391,1112 @@ int benchFiles(const char** fileNamesTable, int nbFiles)
+    g_winner.result.cSize = (size_t)-1;
     /* Init */
-    if (inFile==NULL) { DISPLAY( "Pb opening %s\n", inFileName); return 11; }
-    if (inFileSize == UTIL_FILESIZE_UNKNOWN) {
-        DISPLAY("Pb evaluatin size of %s \n", inFileName);
-        fclose(inFile);
-        return 11;
+    if(!cParamValid(paramTarget)) {
+        return 10;
     }
 
-    /* Memory allocation & restrictions */
-    if ((U64)benchedSize > inFileSize) benchedSize = (size_t)inFileSize;
-    if (benchedSize < inFileSize) {
-        DISPLAY("Not enough memory for '%s' \n", inFileName);
-        fclose(inFile);
-        return 11;
-    }
+    /* load dictionary*/
+    if (dictFileName != NULL) {
+        U64 const dictFileSize = UTIL_getFileSize(dictFileName);
+        if (dictFileSize > 64 MB) {
+            DISPLAY("dictionary file %s too large", dictFileName);
+            ret = 10;
+            goto _cleanUp;
+        }
+        dictBufferSize = (size_t)dictFileSize;
+        dictBuffer = malloc(dictBufferSize);
+        if (dictBuffer==NULL) {
+            DISPLAY("not enough memory for dictionary (%u bytes)",
+                            (U32)dictBufferSize);
+            ret = 11;
+            goto _cleanUp;
+
+        }
 
-    /* Alloc */
-    origBuff = malloc(benchedSize);
-    if(!origBuff) {
-        DISPLAY("\nError: not enough memory!\n");
-        fclose(inFile);
-        return 12;
+        {
+            int errorCode = BMK_loadFiles(dictBuffer, dictBufferSize, &dictBufferSize, &dictFileName, 1);
+            if(errorCode) {
+                ret = errorCode;
+                goto _cleanUp;
+            }
+        }
     }
 
     /* Fill input buffer */
-    DISPLAY("Loading %s...       \r", inFileName);
-    {   size_t const readSize = fread(origBuff, 1, benchedSize, inFile);
-        fclose(inFile);
-        if(readSize != benchedSize) {
-            DISPLAY("\nError: problem reading file '%s' !!    \n", inFileName);
-            free(origBuff);
-            return 13;
-    }   }
+    if(nbFiles == 1) {
+        DISPLAY("Loading %s...       \r", fileNamesTable[0]);
+    } else {
+        DISPLAY("Loading %lu Files...       \r", (unsigned long)nbFiles); 
+    }
+
+    {   
+        U64 const totalSizeToLoad = UTIL_getTotalFileSize(fileNamesTable, (U32)nbFiles);
+        int ec;
+        unsigned i;
+        benchedSize = BMK_findMaxMem(totalSizeToLoad * 3) / 3;
+
+        origBuff = malloc(benchedSize);
+        if(!origBuff || !fileSizes) {
+            DISPLAY("Not enough memory for stuff\n");
+            ret = 1;
+            goto _cleanUp;
+        }
+        ec = BMK_loadFiles(origBuff, benchedSize, fileSizes, fileNamesTable, (U32)nbFiles);
+        if(ec) {
+            DISPLAY("Error Loading Files");
+            ret = ec;
+            goto _cleanUp;
+        }
+        benchedSize = 0;
+        for(i = 0; i < nbFiles; i++) {
+            benchedSize += fileSizes[i];
+        }
+        origBuff = realloc(origBuff, benchedSize);
```

Comment:
the usage of `realloc()` can be pretty dangerous.
If it fails to allocate a new buffer of size `benchedSize`, it will actually return `NULL`,
but the initial buffer will still be allocated.
By overwriting `origBuff`, you lose track of the original buffer, making it impossible to `free()`.

You should also test the result of `realloc()`, since success is not guaranteed.